text/0000-build-plan-merging.md

```diff
@@ -0,0 +1,260 @@
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Build Plan (TOML) Changes
+We propose a complete restructuring of the `requires` field and an addition to the `provides` field in the Build Plan (TOML) file.
+
+We propose the following Build Plan (TOML):
+
+```toml
+[[provides]]
+name = "<dependency name>"
+[provides.strategy.version]
+  collect = false
+
+[[requires]]
+name = "<dependency name>"
+[requires.version]
+  constraint = "<dependency constraint>"
+  source = "<dependency constraint source>"
+[requires.capabilities]
+  # arbitary keys to bools
+  build = false
+  launch = false
+[requires.modifiers]
+  # arbitrary keys and values
+  some-modifer = "modifier"
+
+[[or]]
+
+[[or.provides]]
+name = "<dependency name>"
+[provides.strategy.version]
+  collect = false
+
+[[or.requires]]
+name = "<dependency name>"
+[requires.version]
+  constraint = "<other dependency constraint>"
+  source = "<other dependency constraint source>"
+[requires.capabilities]
+  # arbitary keys to bools
+  build = false
+  launch = false
+[requires.modifiers]
+  # arbitrary keys and values
+  some-other-modifer = "other modifier"
+```
+
+This proposal adds a new field to `provides`:
+- `provides.strategy.version` field holds a `collect` key which indicates if the version fields from all of the requires should be merged together or if they should collected into an array.
+  - This field is structured to be extensible to other fields in `requires`.
+
+This proposal also removes both the `requires.version` and `requires.metadata` fields and replaces them with three new fields each with their own purpose:
+- `requires.version` field holds all of the version requirement information.
+  - `constraint` can be any semver constraint used to decide which dependency the providing buildpack will install.
+  - `source` is the origin of the constraint (i.e. it came from `package.json`).
+- `requires.capabilities` holds the `build` and `launch` keys and indicate when during which phase buildpack specific data and dependencies are available.
+- `requires.modifiers` is meant to hold only specific and **unique** buildpack communications.
+
+
+## Buildpack Plan (TOML) changes
+For the same reasons, we would like to add the `version`, `capabilities`, and `modifiers` fields to the Buildpack Plan, along with a `versions` field and some additional format changes:
+
+```toml
+[[entries]]
+name = "<dependency name>"
+[entries.version]
+  constraint = "<dependency constraint>"
+[entries.capabilities]
+  # or merged bool values from arbitray keys
+  build = false
+  launch = false
+[entries.modifiers]
+  # last-in-wins merged values from arbitrary keys
+  some-modifer = "modifier"
+[[entries.collection]]
+[entries.collection.version]
+  constraint = "<dependency constraint>"
+  source = "<dependency constraint source>"
+```
+In each entry the `name` field is **unique**.
+
+We arrived at this format for several reasons:
+- Moves some common buildpack operations into the lifecycle.
```

Comment:
Why is the lifecycle the right place for this vs. the language bindings or helper libraries? If the lifecycle merely provides _all_ of the data, a helper library can give present it in a merged on unmerged format on demand, without requiring that preference to move into the spec (as `provides.strategy`).